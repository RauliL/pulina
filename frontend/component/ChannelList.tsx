import * as React from 'react';
import { Button } from 'reactstrap';

import { Channel } from '../types/channel';

import ChannelListItem from './ChannelListItem';

export interface StateProps {
  channels: Channel[];
  currentChannel?: string;
}

export interface DispatchProps {
  onSelectChannel: (channel: string) => void;
  onToggleChannelList: () => void;
}

export type Props = StateProps & DispatchProps;

const ChannelList: React.SFC<Props> = (props) => (
  <ul className="list-unstyled">
    <li className="d-lg-none navbar bg-white text-body border-bottom">
      <span className="navbar-brand">Channels</span>
      <div className="float-right">
        <Button outline={true} onClick={props.onToggleChannelList}>
          <i className="fas fa-times"/>
        </Button>
      </div>
    </li>
    {props.channels.map((channel) => (
      <ChannelListItem
        key={channel.name}
        name={channel.name}
        onSelect={() => props.onSelectChannel(channel.name)}
        isActive={props.currentChannel === channel.name}
        hasUnreadMessages={channel.hasUnreadMessages}
        hasUnreadHighlight={channel.hasUnreadHighlight}
      />
    ))}
  </ul>
);

export default ChannelList;
