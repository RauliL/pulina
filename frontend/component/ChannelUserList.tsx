import * as React from 'react';

export interface StateProps {
  users?: string[];
}

export type Props = StateProps;

const ChannelUserList: React.SFC<Props> = ({ users }) => (
  <ul className="list-unstyled" style={{ overflowY: 'auto' }}>
    {users && users.sort().map((user) => (
      <li className="p-2 border-bottom" key={user}>
        {user}
      </li>
    ))}
  </ul>
);

export default ChannelUserList;
