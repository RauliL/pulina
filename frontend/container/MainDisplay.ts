import { connect } from 'react-redux';

import { onChannelError, toggleChannelList, toggleUserList } from '../action';
import MainDisplay, {
  DispatchProps,
  StateProps,
} from '../component/MainDisplay';
import { PulinaDispatch, PulinaState } from '../types';
import { getCurrentChannel } from '../utils';

const mapStateToProps = (state: PulinaState): StateProps => ({
  isNickRegistered: state.nick != null,
  currentChannel: getCurrentChannel(state),
});

const mapDispatchToProps = (dispatch: PulinaDispatch): DispatchProps => ({
  onCommandError(errorMessage: string) {
    dispatch(onChannelError(errorMessage));
  },

  onToggleChannelList() {
    dispatch(toggleChannelList());
  },

  onToggleUserList() {
    dispatch(toggleUserList());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(MainDisplay);
