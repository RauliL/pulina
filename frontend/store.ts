import { createStore } from 'redux';

import { reducer } from './reducer';
import { Action, State, Store } from './types/store';

export const initializeStore = (): Store => createStore<State, Action, {}, {}>(
  reducer,
  process.env.NODE_ENV === 'development' &&
  window &&
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__() :
  undefined,
);
