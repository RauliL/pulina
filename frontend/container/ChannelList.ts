import { values } from 'lodash';
import { connect } from 'react-redux';

import { changeCurrentChannel } from '../action';
import ChannelList, {
  DispatchProps,
  StateProps,
} from '../component/ChannelList';
import { PulinaDispatch, PulinaState } from '../types';

const mapStateToProps = (state: PulinaState): StateProps => ({
  channels: values(state.channels),
  currentChannel: state.currentChannel,
});

const mapDispatchToProps = (dispatch: PulinaDispatch): DispatchProps => ({
  onSelectChannel(channel: string) {
    dispatch(changeCurrentChannel(channel));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ChannelList);
