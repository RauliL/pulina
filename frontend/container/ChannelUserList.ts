import { connect } from 'react-redux';

import { toggleUserList } from '../action';
import ChannelUserList, {
  DispatchProps,
  StateProps,
} from '../component/ChannelUserList';
import { PulinaDispatch, PulinaState } from '../types';
import { getCurrentChannel } from '../utils';

const mapStateToProps = (state: PulinaState): StateProps => {
  const currentChannel = getCurrentChannel(state);
  let channelName: string | undefined;
  let users: string[] | undefined;

  if (currentChannel) {
    channelName = currentChannel.name;
    users = currentChannel.users.sort();
  } else {
    users = [];
  }

  return { channelName, users };
};

const mapDispatchToProps = (dispatch: PulinaDispatch): DispatchProps => ({
  onToggleUserList() {
    dispatch(toggleUserList());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ChannelUserList);
