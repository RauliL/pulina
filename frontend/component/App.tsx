import * as React from 'react';
import { Col, Container, Row } from 'reactstrap';

import { ClientCommand } from '../../common/command';

import ChannelList from '../container/ChannelList';
import ChannelUserList from '../container/ChannelUserList';
import MainDisplay from '../container/MainDisplay';

export interface Props {
  socket: Socket;
}

export default class App extends React.Component<Props> {
  public componentWillUnmount() {
    this.props.socket.close();
  }

  public render() {
    return (
      <Container fluid={true}>
        <Row>
          <Col
            lg={1}
            className="bg-dark p-0 border-right"
            style={{
              height: '100vh',
              overflowY: 'auto',
            }}
          >
            <ChannelList/>
          </Col>
          <Col lg={10} className="p-0" style={{ height: '100vh' }}>
            <MainDisplay onCommand={this.onCommand}/>
          </Col>
          <Col
            lg={1}
            className="p-0 border-left"
            style={{
              height: '100vh',
              overflowY: 'auto',
            }}>
            <ChannelUserList/>
          </Col>
        </Row>
      </Container>
    );
  }

  private onCommand = <T extends ClientCommand> (command: T) => {
    this.props.socket.emit(command.type, command);
  };
}
