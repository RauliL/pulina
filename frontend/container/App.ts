import { connect } from 'react-redux';

import App, { StateProps } from '../component/App';
import { PulinaState } from '../types';

const mapStateToProps = (state: PulinaState): StateProps => ({
  channelListVisible: state.channelListVisible,
  userListVisible: state.userListVisible,
});

export default connect(mapStateToProps)(App);
