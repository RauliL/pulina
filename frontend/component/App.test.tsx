import { mount, shallow } from 'enzyme';
import { defaultTo, noop } from 'lodash';
import * as React from 'react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';

import ChannelList from '../container/ChannelList';
import ChannelUserList from '../container/ChannelUserList';
import MainDisplay from '../container/MainDisplay';
import { State } from '../types/store';

import App, { Props } from './App';

describe('<App/>', () => {
  const mockStore = createMockStore<State>();
  const renderComponent = (props: Partial<Props> = {}) => (
    <Provider store={mockStore()}>
      <App
        channelListVisible={defaultTo(props.channelListVisible, false)}
        userListVisible={defaultTo(props.userListVisible, false)}
        socket={{
          on: noop,
          emit: noop,
          close: noop,
        }}
      />
    </Provider>
  );
  const mountComponent = (props: Partial<Props> = {}) => mount(
    renderComponent(props),
  );

  it('should render without throwing exception', () => {
    expect(shallow(renderComponent())).toMatchSnapshot();
  });

  it('should render <ChannelList/>', () => {
    expect(mountComponent().find(ChannelList)).toHaveLength(1);
  });

  it('should render <MainDisplay/>', () => {
    expect(mountComponent().find(MainDisplay)).toHaveLength(1);
  });

  it('should render <ChannelUserList/>', () => {
    expect(mountComponent().find(ChannelUserList)).toHaveLength(1);
  });
});
