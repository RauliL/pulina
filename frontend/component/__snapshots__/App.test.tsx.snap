// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> should render without throwing exception 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <App
    channelListVisible={false}
    socket={
      Object {
        "close": [Function],
        "emit": [Function],
        "on": [Function],
      }
    }
    userListVisible={false}
  />
</ContextProvider>
`;
